### ISSUE TITLE
Funny performance in gradients

### ISSUE THREAD

**Issue by Dapid** [google/jax#934]
[Reproducing code][1]. When I run a JIT function several times, the first time it needs to compile it is slow, but then becomes much faster, as expected. Here is a plot of the timings for successive calls for different batch sizes:

![iterations](https://user-images.githubusercontent.com/1616942/60276156-60d07c00-98fb-11e9-89a5-6d615396843e.png)

What is going on? It is super-fast for three iterations, and then slows down by two orders of magnitude! I don't think it has to do with memory issues, since it takes the exact same number of iterations for batches three times bigger.

The late times, after the performance is stable, make sense. Every iteration takes longer, but the time per sample is faster:
![average](https://user-images.githubusercontent.com/1616942/60276588-33d09900-98fc-11e9-9c01-ac493cf72d64.png)

On the other hand, the minimum time seems to be faster the more data is there (but it is noisy)

![min_time](https://user-images.githubusercontent.com/1616942/60276838-ac375a00-98fc-11e9-8073-5045637c16e4.png)

This only happens on the GPU, the  "official" JAX examples also behave normally, as well as this example on the CPU:

![iterations_cpu](https://user-images.githubusercontent.com/1616942/60277534-0edd2580-98fe-11e9-9ee7-96517469465d.png)

As a note: it seems that CPU is still faster the larger the batch:

![average_cpu](https://user-images.githubusercontent.com/1616942/60277341-a68e4400-98fd-11e9-96e7-a8ec0292aa26.png)

This is wall time per batch.

   [1]: https://gist.github.com/Dapid/80751f98ec5c51c926bd9c4171ced282

**Comment by hawkinsp** [comment]
Does https://jax.readthedocs.io/en/latest/async_dispatch.html explain the issue?

**Comment by Dapid** [comment]
Ah, yes. Blocking I get reasonable results. Now GPU is faster than the CPU, and scales better, and the bigger the batch the longer it takes.
